framework/ChiMesh/SweepUtilities/SweepBoundary/sweep_boundaries.h

```diff
@@ -110,39 +136,117 @@ class BoundaryReflecting : public BoundaryBase
+                     chi_math::CoordinateSystemType::CARTESIAN) :
+    BoundaryBase(BoundaryType::REFLECTING,in_num_groups,coord_type),
+    normal_(in_normal)
   {}
 
-  double* HeterogenousPsiIncoming(
-                          int angle_num,
-                          uint64_t cell_local_id,
-                          int face_num,
-                          int fi,
-                          int gs_ss_begin) override;
-  double* HeterogenousPsiOutgoing(
-                          int angle_num,
-                          uint64_t cell_local_id,
-                          int face_num,
-                          int fi,
-                          int gs_ss_begin) override;
+  const chi_mesh::Vector3& Normal() const {return normal_;}
+  bool IsOpposingReflected() const {return opposing_reflected_;}
+  void SetOpposingReflected(bool value) { opposing_reflected_ = value;}
+
+  std::vector<AngVec>& GetHeteroBoundaryFluxNew() {return hetero_boundary_flux_;}
+  std::vector<AngVec>& GetHeteroBoundaryFluxOld() {return hetero_boundary_flux_old_;}
+
+  std::vector<int>& GetReflectedAngleIndexMap() {return reflected_anglenum_;}
+  std::vector<std::vector<bool>>&
+  GetAngleReadyFlags() {return angle_readyflags_;}
+
+  double* HeterogenousPsiIncoming(uint64_t cell_local_id,
+                                  int face_num,
+                                  int fi,
+                                  int angle_num,
+                                  int group_num,
+                                  int gs_ss_begin) override;
+  double* HeterogenousPsiOutgoing(uint64_t cell_local_id,
+                                  int face_num,
+                                  int fi,
+                                  int angle_num,
+                                  int gs_ss_begin) override;
 
   void UpdateAnglesReadyStatus(const std::vector<size_t>& angles,
                                size_t gs_ss) override;
   bool CheckAnglesReadyStatus(const std::vector<size_t>& angles,
                               size_t gs_ss) override;
   void ResetAnglesReadyStatus();
 };
+
+/**This boundary function class can be derived from to
+ * provide a much more custom experience. This function
+ * is called during Setup. */
+class BoundaryFunction
+{
+public:
+  virtual std::vector<double> Evaluate(
+    size_t cell_global_id,
+    int    cell_material_id,
+    unsigned int face_index,
+    unsigned int face_node_index,
+    const chi_mesh::Vector3& face_node_location,
+    const chi_mesh::Vector3& face_node_normal,
+    const std::vector<int>& quadrature_angle_indices,
+    const std::vector<chi_mesh::Vector3>& quadrature_angle_vectors,
+    const std::vector<std::pair<double,double>>& quadrature_phi_theta_angles,
+    const std::vector<int>& group_indices)
+  {
+    size_t num_angles = quadrature_angle_indices.size();
+    size_t num_groups = group_indices.size();
+
+    std::vector<double> psi(num_angles * num_groups, 0.0);
+
+    return psi;
+  }
```

Comment:
We should add a default argument `const double time = 0.0` to this to handle transients.


Is there a reason for returning zeros? Why not make this abstract?

Reply:
Two very good points